src/Microsoft.AspNetCore.SignalR.Client/HubConnectionBuilder.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel;
+using System.Net.Http;
+using Microsoft.AspNetCore.SignalR.Internal.Protocol;
+using Microsoft.AspNetCore.Sockets;
+using Microsoft.AspNetCore.Sockets.Client;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.SignalR.Client
+{
+    public class HubConnectionBuilder
+    {
+        private readonly Uri _url;
+        private ILoggerFactory _loggerFactory;
+        private TransportType _transportType = TransportType.All;
+        private IHubProtocol _hubProtocol;
+        private HttpMessageHandler _httpMessageHandler;
+
+        public HubConnectionBuilder(Uri url)
+        {
+            _url = url;
+        }
+
+        public HubConnectionBuilder WithLogger(ILoggerFactory loggerFactory)
+        {
+            _loggerFactory = loggerFactory;
+            return this;
+        }
+
+        public HubConnectionBuilder WithTransportType(TransportType transportType)
+        {
+            _transportType = transportType;
+            return this;
+        }
+
+        public HubConnectionBuilder WithHubProtocol(IHubProtocol hubProtocol)
+        {
+            _hubProtocol = hubProtocol;
+            return this;
+        }
+
+        public HubConnectionBuilder WithHttpMessageHandler(HttpMessageHandler httpMessageHandler)
+        {
+            _httpMessageHandler = httpMessageHandler;
+            return this;
+        }
+
+        public HubConnection Build()
+        {
+            var httpConnection = new HttpConnection(_url, _transportType, _loggerFactory, _httpMessageHandler);
+            return new HubConnection(httpConnection, _hubProtocol ?? new JsonHubProtocol(new JsonSerializer()), _loggerFactory);
```

Comment:
Should we push this default hubProtocol logic into the HubConnection class itself?

Reply:
I think it makes no difference.

Reply:
I'd actually prefer the default live inside the builder where it's easier to configure, and it keeps the `HubConnection` type pure.

Reply:
Should we remove the overloads from `HubConnection` now that we have the builder, to avoid duplicate code that could drift out of sync? `HubConnection` should just expect to be provided everything it needs now, right? (and throw `ArgumentNullException` for missing arguments)

Reply:
I think we should leave the overloads on HubConnection. Unless we basically want to force people to use the builder

Reply:
I just removed the overloads :) We want people to use the builder.